### DISCUSSION TITLE
Escaping dollar sign with uppercase in fix string

### DISCUSSION THREAD

**Discussion by kam193** [ast-grep/ast-grep#1322] [Q&A] [answered]
Hey!

I've started playing around ast-grep, and I find it awesome. However, I feel like I've found an edge case that is extremely painful for me: I cannot find a way to properly escape `$` in `fix` field. My use case requires rewriting strings that can contain dollar sign followed by upper case char, as in this playground:

https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6InB5dGhvbiIsInF1ZXJ5IjoicideW0EtWmEtejAtOV8tXSsnIiwicmV3cml0ZSI6IiIsImNvbmZpZyI6InJ1bGU6XG4gICAgcGF0dGVybjogdmFyaWFibGUgPSBcInZhbHVlXCJcbmZpeDogdmFyaWFibGUgPSBcIlNvbWUkVmFsdWVXaXRoRG9sbGFyXCIiLCJzb3VyY2UiOiJ2YXJpYWJsZSA9IFwidmFsdWVcIiJ9

Don't I just see the proper method, or is it currently impossible to prevent metavariable being evaluated? Playground throws an error in any escaping syntax I can imagine; when I work with `sg run` (with proper shell escaping or even without shell being involved), the tool just removes metavariable-like parts. 

Do you have any idea for a solution?

**Comment by HerringtonDarkholme** [comment, marked as answer]
Hi, thanks for using ast-grep.
Escaping dollar sign is unfortunately not well-supported now.

**Reply by kam193** [reply]
Thanks for the answer!

**Reply by HerringtonDarkholme** [reply]
One work around is using JS/Py API, but I don't think one should bother with this for just rewriting with `$` value. 

Another work around is using two meta-variables and `replace` transformation. https://ast-grep.github.io/reference/yaml/transformation.html#replace

https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6InB5dGhvbiIsInF1ZXJ5IjoicideW0EtWmEtejAtOV8tXSsnIiwicmV3cml0ZSI6IiIsImNvbmZpZyI6InJ1bGU6XG4gICAgcGF0dGVybjogdmFyaWFibGUgPSBcIiRWQUxVRVwiXG50cmFuc2Zvcm06XG4gICAgRklSU1Q6XG4gICAgICAgIHJlcGxhY2U6XG4gICAgICAgICAgICBieTogJ1NvbWUkJ1xuICAgICAgICAgICAgcmVwbGFjZTogJy4rJ1xuICAgICAgICAgICAgc291cmNlOiAkVkFMVUVcbiAgICBPVEhFUjpcbiAgICAgICAgcmVwbGFjZTpcbiAgICAgICAgICAgIGJ5OiBWYWx1ZVdpdGhEb2xsYXJcbiAgICAgICAgICAgIHJlcGxhY2U6IC4rXG4gICAgICAgICAgICBzb3VyY2U6ICRWQUxVRVxuZml4OiB2YXJpYWJsZSA9IFwiJEZJUlNUJE9USEVSXCIiLCJzb3VyY2UiOiJ2YXJpYWJsZSA9IFwidmFsdWVcIiJ9

The trick here is split the `$` value into two strings and concate them together.

<img width="256" alt="image" src="https://github.com/user-attachments/assets/2dfeab9f-0ed8-46dc-bf44-9089af1528bd">

**Reply by kam193** [reply]
This is an interesting idea, I may go with this eventually. So far I use the extremely basic workaround: I replace (outside the last-grep) dollar with some text, than do the fix, and then replace it back in the final file. It could be better, but works alright.

**Reply by kam193** [reply]
The Python API could be a solution, because in this case I generate the rule dynamically. However, so far it worked better to just call CLI. Especially that I mix pre-prepared and dynamically generated rules, so to use Python API I would eventually need to re-read file a few times. Native ast-grep call does it better
